### ISSUE TITLE
Questions about SPAdes scaffold and multi-library

### ISSUE THREAD

**Issue by liaoyunshi** [ablab/spades#126]
Hi,

Thanks for creating such wonderful software.

1. I just used SPAdes to assemble my illumina PE 150 data, and find there are some Ns in the middle of a resulting scaffold. So I wonder how SPAdes create scaffold from contigs. Does it use the position of paired-end reads in different contigs to resolve the scaffold? If so, how can SPAdes get the insert size of paired-end reads as I do not input any parameter about this. Using the position of paired-end reads in the same contigs to estimate the distribution of insert size?

After that, I got many scaffolds but no one reach the predicted size of my products, so I want to do further scaffolding based on overlapping relationship among the current scaffold sequences. Do you have any suggestion about tools doing that?

2. According to manual, I find we can specify multiple (up to 9) libraries in SPAdes via `--pe<#>-1 /-2` options, I am not sure about the meaning of multiple library here. Is that mean different libraries from the same sample or different libraries from different samples. I guess it is the former one which can handle multi libraries from the only one sample but with different setting (e.g. insert size) and generate final assembly for the one sample?

3. By default, the k-mer set is 21,33,55, but in what situation I should adjust the setting and how to change it according to different libraries?

4. SPAdes can correct reads based on clustering, which I would think revise error-prone bases, but it seems after correction, the total read number decrease and some reads even become single read, so is it any read deletion during correction, any criteria to decide delete or not?

Thanks for your help. Below please find the sample command I used.

Command line: 
`xxx/SPAdes-3.12.0-Darwin/bin/spades.py -t 4 --careful -1 Barcode01_1_trim35_qq.fq -2 Barcode01_2_trim35_qq.fq -o spades_output_Barcode01`

**Comment by liaoyunshi** [comment]
> Normally it does not make any sense to scaffold more since it's the insert size distribution and the repeat content influences the scaffold lengths.

I agree that. But actually I want to assemble a small genome, so I would hope to see more integrated contigs. However, the number of resulting contigs is not quite few, so I try to use some software to seek the potential overlap among the resulting scaffolds. And I do find some overlap, for example (the number of overlapping bases and flanking regions is just an example):
        
nnnnnnnnnAATCATCGATAT
                  AATCATCGATATnnnnnnnnnnn
or
nnnnnnnnnAATCAT**T**GATAT
                  AATCAT**C**GATATnnnnnnnnnnn

I just wonder why SPAdes cannot merge this two fully/mostly overlapping contigs together (especially for the fully overlapped one), is it about the length of ovelap and the length of K-mer?

> The SPAdes output does not depend on the amount of memory, though it may depends on the # of threads used.

So I wonder the # of threads would affect the speed of assembly or even the whole assembly process and lead to the differences of resulting contigs and scaffolds? If the latter is right, I am quite worried about the  consequence of few # of threads. And may I know if there is minimum  # of threads needed for SPAdes?

Thanks.

**Comment by asl** [comment]
> And I do find some overlap, for example (the number of overlapping bases and flanking regions is just an example):

Sure. The consecutive contigs / scaffolds do overlap on the k-mer. The reason why SPAdes does not "merge" them is that there is no unambiguous way to resolve this repeat. You could visualize the assembly graph to figure out the graph structure there.

> And may I know if there is minimum # of threads needed for SPAdes?

The minimum number is 1.

> So I wonder the # of threads would affect the speed of assembly

Correct

> or even the whole assembly process and lead to the differences of resulting contigs and scaffolds? I

It may in general yield slightly different assemblies.

**Comment by liaoyunshi** [comment]
Thanks. And I still have one last issue, that is the warning in the end of log file.

> === Error correction and assembling warnings:
>  * 0:00:02.200    16M / 3G    WARN    General                 (kmer_coverage_model.cpp   : 218)   Too many erroneous kmers, the estimates might be unreliable
>  * 0:00:09.868    16M / 3G    WARN    General                 (kmer_coverage_model.cpp   : 327)   Valley value was estimated improperly, reset to 1
>  * 0:00:09.869    16M / 3G    WARN    General                 (kmer_coverage_model.cpp   : 366)   Failed to determine erroneous kmer threshold. Threshold set to: 1
> ======= Warnings saved to xxx

May I know what is the problem behind this warning, would it affect the accuracy of resulting contigs?

Thanks.

**Comment by asl** [comment]
It may affect. The warning indicates that there might be some unexpected non-uniformity in the coverage.
